### ISSUE TITLE
Dynamic URLs do not work anymore with journey 1.0.2 

### ISSUE THREAD

**Issue by etcs** [rails/journey#35]
I upgraded from rails 3.2.1 to 3.2.3 and dynamic URLs did not work anymore. The problem was related to the upgrade of journey 1.0.1 to journey 1.0.3, but also the downgrade to journey 1.0.2 did not help; 1.0.2 has the same problem. 1.0.1 works fine.
The failure  is caused by line 107 in formatter.rb:
journey 1.0.1: true
journey 1.0.2: parts.fetch(key) {false}

It is also the only difference between both files, so If I use journey 1.0.2 and if I replace formatter.rb from that one from journey 1.0.1, everything works fine.

In the example below you'll see, that the failure message does not match the request. The request should be routed to the action 'show_cluster', but it is routed to 'concrete_testcase', which comes in alphabetical order before 'show_cluster' ...  

<b>URL:</b>
http://192.168.123.199:3000/test_cluster/show_cluster/93

<b>failure message in browser: </b>
Routing Error

No route matches {:controller=>"test_cluster", :action=>"concrete_testcase"}

Try running rake routes for more information on available routes. 

<b>rake routes: </b>
<pre><code>test_cluster_show_cluster GET  /test_cluster/show_cluster/:rel_id(.:format)                         test_cluster#show_cluster
test_cluster_concrete_testcase GET  /test_cluster/concrete_testcase/:rel_id/:cluster_id(.:format)        test_cluster#concrete_testcase</pre></code>

<b>routes.rb: </b>
<pre><code>get "test_cluster/show_cluster/:rel_id" => "test_cluster#show_cluster", as: "test_cluster_show_cluster"
get "test_cluster/concrete_testcase/:rel_id/:cluster_id" => "test_cluster#concrete_testcase", as: "test_cluster_concrete_testcase"</pre></code>

**Comment by pixeltrix** [comment]
It's not a recognition problem, it's a generation problem - somewhere on that page you're using `test_cluster_concrete_testcase_path` with a `nil` for `:cluster_id`. This may be due to passing a new model instance which returns `nil` for `to_param`.

Duplicate of #20.

**Comment by etcs** [comment]
You are right, but it was not a mistake.I thought, it is some kind of 'rails magic', that I simply could use test_cluster_concrete_testcase_path in show_cluster.html.erb, which was modified using :rel_id of the parent page (in journey 1.0.1). The :rel_id was taken from "test_cluster/show_cluster/:rel_id" path and used by rails to build  "test_cluster/concrete_testcase/:rel_id" path. Great!

In other words: the semantics of the *_path methods have been changed between journey 1.0.1 and journey 1.0.2

<b>journey 1.0.1: show_cluster.html.erb</b>
<code>'print' <% test_cluster_concrete_testcase_path=<%= test_cluster_concrete_testcase_path %>
<%= link_to 'CTC', test_cluster_concrete_testcase_path + '/'+ my_id.to_s %> </code>

output:
test_cluster_concrete_testcase_path=/test_cluster/concrete_testcase/93
link:  ip+port/test_cluster/concrete_testcase/93/my_id

<b>journey 1.0.2: show_cluster.html.erb</b>
<code>'print'<% test_cluster_concrete_testcase_path=<%= test_cluster_concrete_testcase_path %> 
<%= link_to 'CTC', test_cluster_concrete_testcase_path(:rel_id => rel_id.to_s, :cluster_id => my_id.to_s) %></code>

output:
first line: not possible (routing failure)
link:  ip+port/test_cluster/concrete_testcase/93/my_id

<b>conclusion</b>
I do not know the reason for the change, but I tend to prefer semantics of journey 1.0.1. I see following advantages:
- the _path-methods can be used without parameters. If there are variable names in routes.rb, which are defined in both current and parent page, rails expands path using the corresponding variable values of the parent page in the current page.
- there is no need to use names of variables of routes.rb in erb.html files, the path can simply be built by *_path + '/' + my_id (It is not possible with journey 1.0.2). This leads to less dependencies between several files of code (have a look at the second line of the code examples;  both do the same, but the variant of  journey 1.0.2 is much more complicated)

**Comment by pixeltrix** [comment]
When using positional arguments to the route helpers they always need to be specified in full - the change in Journey 1.0.2 was to fix a regression against Rails 3.1 in Rails 3.2. You can use the recall parameters using hash parameters though - this should work:

``` erb
<%= link_to 'CTC', test_cluster_concrete_testcase_path(:cluster_id => my_id) %>
```

You don't need to use `to_s` on the id - the values are converted to strings using `to_param` inside the route helpers.
